gradio/blocks.py

```diff
@@ -1071,11 +1071,16 @@ async def call_function(
             else:
                 fn = block_fn.fn
 
+            def fn_wrap(*args, **kwargs):
+                queueing.thread_data.blocks = self
+                queueing.thread_data.event_id = event_id
+                return fn(*args, **kwargs)
+
             if inspect.iscoroutinefunction(fn):
                 prediction = await fn(*processed_input)
             else:
                 prediction = await anyio.to_thread.run_sync(
-                    fn, *processed_input, limiter=self.limiter
+                    fn_wrap, *processed_input, limiter=self.limiter
```

Comment:
This means that this will only work for regular functions, not async functions or generators (?)

Reply:
Why do we need to do this? Is it so that the queue knows which blocks corresponds to the function it is currently processing? I don't think this is necessary -- we don't need it to handle generators for example

Reply:
I think you can handle `gr.Info` and `gr.Warning` in the same way generators are handled -- see https://github.com/gradio-app/gradio/blob/d65512cb3a91e2c5b6abf70b52f56a51195c47e7/gradio/queueing.py#L367

Reply:
This is needed because there needs to be some communication between the gr.Warning() and the Gradio server. Since the gr.Warning is not returned, the server does not know if the gr.Warning is issued. Therefore, the gr.Warning needs to have a reference to the blocks server and the event id that issued it. I do this by storing references to both within the thread as thread locals. `anyio.to_thread` spins off a new thread, so I need to put this in another function called `fn_wrap`.

This works fine with generators and async. Try here:
```python
import gradio as gr
import time

def greet(name):
    gr.Info("Starting...")
    time.sleep(2)
    for i, letter in enumerate(name):
        if letter == " ":
            gr.Warning("Space detected!")
        yield "Hello " + name[:i+1] + "!"
        time.sleep(0.5)
    time.sleep(2)
    raise gr.Error("Error!")

demo = gr.Interface(fn=greet, inputs="text", outputs="text")
demo.queue().launch() 
```

Reply:
Will take a look today/tomorrow @aliabid94 ! I can confirm generators work in your use case, but warnings are not displayed with `every`

```python
import gradio as gr


def warn_1():
    gr.Warning("Warning 1!")
    return "foo"

def warn_2():
    gr.Warning("Warning 2!")
    return "bar"

with gr.Blocks() as demo:
    t1 = gr.Textbox()
    t2 = gr.Textbox()

    button = gr.Button("Warning 1")
    button.click(warn_1, None, t1, every=1)

    button2 = gr.Button("Warning 2")
    button2.click(warn_2, None, t2)
    

demo.queue().launch()
```

Reply:
fixed